rubix-core/src/main/java/com/qubole/rubix/core/CachingInputStream.java

```diff
@@ -384,37 +394,75 @@ else if (isCached.get(idx).getLocation() == Location.CACHED) {
             else {
                 if (isCached.get(idx).getLocation() == Location.NON_LOCAL) {
                     String remoteLocation = isCached.get(idx).getRemoteLocation();
-                    log.debug(String.format("Sending block %d to NonLocalReadRequestChain to node : %s", blockNum, remoteLocation));
-                    if (!nonLocalRequests.containsKey(remoteLocation)) {
-                        NonLocalReadRequestChain nonLocalReadRequestChain =
-                            new NonLocalReadRequestChain(remoteLocation, fileSize, lastModified, conf,
-                                remoteFileSystem, remotePath, clusterType.ordinal(), strictMode, statistics);
-                        nonLocalRequests.put(remoteLocation, nonLocalReadRequestChain);
+
+                    if (CacheConfig.isParallelWarmupEnabled(conf)) {
+                        log.info(String.format("Sending block %d to NonLocalRequestChain to node : %s", blockNum, remoteLocation));
+                        if (!nonLocalAsyncRequests.containsKey(remoteLocation)) {
+                            NonLocalRequestChain nonLocalRequestChain =
+                                new NonLocalRequestChain(remoteLocation, fileSize, lastModified,
+                                  conf, remoteFileSystem, remotePath, clusterType.ordinal(), strictMode,
+                                  statistics, nextReadBlock, endBlock);
+                            nonLocalAsyncRequests.put(remoteLocation, nonLocalRequestChain);
+                        }
+                        nonLocalAsyncRequests.get(remoteLocation).addReadRequest(readRequest);
+                        if (!nonLocalAsyncRequests.get(remoteLocation).getDirectReadRequestQueue().isEmpty()) {
+                            if (directReadRequestChain == null) {
+                                nonLocalAsyncRequests.get(remoteLocation).getDirectReadRequestQueue().remove();
```

Comment:
Do we need a DirectReadRequestQueue in NonLocalRequestChain? From this logic it seems there can never be more than one entry in that Queue

Reply:
Removed that queue logic. Also found out one more bug here where the cache status was not returned for the desired block in NonLocalRequestChain. Will be relying on a boolean variable here to figure out whether to need a direct read request or not.